docs/reusing-typecheck-results.md

```diff
@@ -0,0 +1,251 @@
+|>Generated IL and metadata          |  9,2102|  0,0015|    218|      0|      0|      0|    457|     46|
+|>Layout image                       |  9,2195|  0,0025|    218|      0|      0|      0|    457|     46|
+|>Writing Image                      |  9,2272|  0,0006|    218|      0|      0|      0|    456|     46|
+|>Signing Image                      |  9,2338|  0,0000|    218|      0|      0|      0|    456|     46|
+// main6
+|Write .NET Binary                   |  9,2463|  0,1930|    218|      0|      0|      0|    464|     46|     <-- long
+--------------------------------------------------------------------------------------------------------
+```
+
+The phases taking the longest are marked in the right. Those are assembly import, type check, optimization and IL writing.
+
+## Implementation plan
+
+The conclusion from the above is that there is a lot of potential in caching - at the same time, implementing things to the full potential right away will require serious changes in the compiler code, which will be dangerous, hard to test and review. Therefore, I propose to implement the feature in stages where each stage brings gains in some scenarios.
+
+**Stage 1 - force-gen and compare typechecking graphs**
+
+We already create the TC graph in `main1` and we are able to dump it (in Mermaid) via the compiler flags.
+
+So we can do the following:
+- force-gen and save the graph
+- append all the extra compilation info (basically, the compilation arguments string)
+- before doing the retypecheck, detect if there is any change to above:
+  - if so, skip the retypecheck
+  - otherwise, apply some algorithm to get the graph diff. The result should be the list of files for retypecheck (just ignore it for now)
+
+This will allow to skip a huge part of the compilation in the scenarios like reopening Visual Studio for a project when nothing has changed.
+
+**Stage 2 - skip retypecheck for some files**
+
+In `main1`, we do unpickling (= deserializing) and pickling (= serializing) of the code. This currently happens on the module/namespace basis. 
+
+So here we can do the following:
+- extend the pickler logic so that it can work on the file basis
+- save acquired typecheck information to the intermediate folder after the typecheck
+- for the files not needing retypecheck as per graph diff, skip the phase, instead restore the typecheck info
+
+This will hugely benefit the scenarios when only the edge of the graph is affected (think one test in a test project).
+
+**Stage 3 - reduce signature data generation**
+
+In `main3`, we encode signature data in the assemblies.
+
+Since we'll have typecheck graph at this point, we can detect its "edges" - files which nothing depends on. Those don't need to include signature data, hence we can skip the encoding step for them.
+
+Here, the gain largely depends on the type of the project, but this would be anyway a good move in the direction of the compiler less relying on signatures for its operations.
+
```

Comment:
```suggestion
**Stage 3 - reduce signature data generation**

In `main3`, we encode signature data in the assemblies.

Since we are addressing run/debug/test scenarios, a lot of projects can avoid signature data at all. Reason is, signature data is only needed for cross-project compilation. This could be detected automatically for certain types of projects.

For libraries, the resource blob (a byte array) would have to be split into per-file data, and logic added to recombine the splits with freshly typechecked information into a single binary resource back.

```


This does not make much sense.
In order of importance:
- For edge projects (webapp, exe file, test project), signature files can be disable alltogether to save time. This should be the first focus point, since it saves the most and does not need extensive logical change.
- 
- For non-edge project (=libraries), the signature data can be persisted and rehydrated, at the cost of additional logic to make it fit after incremental changes. (i.e. assembling 99/100 pieces if 1 file changed, but then putting it together with 1 re-typechecked file consistently in a single resource block which is still valid and legit)
- The last sentence does not make much sense in the context of this design doc

Reply:
Yeah, I've probably misunderstood this piece during our discussion, addressed now.